flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowSerializer.scala

```diff
@@ -40,15 +46,32 @@ class RowSerializer(val fieldSerializers: Array[TypeSerializer[Any]])
   override def copy(from: Row, reuse: Row): Row = {
     val len = fieldSerializers.length
 
-    if (from.productArity != len) {
+    // cannot reuse, do a non-reuse copy
+    if (reuse == null) {
+      return copy(from)
+    }
+
+    if (from.productArity != len || reuse.productArity != len) {
       throw new RuntimeException("Row arity of reuse and from do not match.")
```

Comment:
Error message should be: "Row arity of reuse or from is incompatible with this RowSerializer".